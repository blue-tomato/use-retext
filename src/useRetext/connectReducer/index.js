import { mapValuesDeep } from '../../helpers';

export default (reducer, emitter, onCall) => {
  mapValuesDeep(reducer, ({ key, value, scope: reducerScope }) => {
    emitter.on(key, ({ scope: emitterScope, payload }) => {
      // Ensure that the reducer-scope starts at least with the emitter-scope
      if (!reducerScope.startsWith(emitterScope)) return;

      onCall({ key, scope: reducerScope, reducer: value, payload });
    });
  });
};
