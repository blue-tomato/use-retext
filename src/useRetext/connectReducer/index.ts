import { Emitter } from 'mitt';
import { mapValuesDeep } from '../../helpers';
import { Reducer, ReducerFuntion } from '../../Store';

interface OnCallObject {
  key: string;
  scope: string;
  reducer: ReducerFuntion;
  payload: any;
}

export default (reducer: Reducer, emitter: Emitter, onCall: (event: OnCallObject) => void) => {
  mapValuesDeep(reducer, ({ key, value, scope: reducerScope }) => {
    emitter.on(key, ({ scope: emitterScope, payload }) => {
      // Ensure that the reducer-scope starts at least with the emitter-scope
      if (!reducerScope.startsWith(emitterScope)) return;

      onCall({ key, scope: reducerScope, reducer: value, payload });
    });
  });
};
