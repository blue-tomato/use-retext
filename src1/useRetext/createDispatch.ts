import mitt from 'mitt';
import { mapValuesDeep } from '../helpers';

export default (action: object): [unknown, mitt.Emitter] => {
  const emitter = mitt();
  const dispatch = mapValuesDeep(action, ({ key, scope }) => (payload?: unknown) =>
    emitter.emit(key, { scope, payload }),
  );

  return [dispatch, emitter];
};
