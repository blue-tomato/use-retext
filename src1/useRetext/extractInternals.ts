import reduce from 'just-reduce-object';
import typeOf from 'just-typeof';
import { Store } from '../Store';

const extractInternals = (store: Store) =>
  reduce(
    store,
    (acc: Array<Record<string, unknown>>, key: string, value: unknown) => {
      const state = acc[0];
      let actions = acc[1];
      let reducer = acc[2];

      if (key === '_actions') {
        // Add to actions
        actions = {
          ...actions,
          ...(value as object),
        };
      } else if (key === '_reducer') {
        // Add to reducer
        reducer = {
          ...reducer,
          ...(value as object),
        };
      } else if (typeOf(value) === 'object') {
        // Extract inner store
        const [extractedState, extractedActions, extractedReducer] = extractInternals(value as Store);

        state[key] = extractedState;

        // Don't add empty objects
        if (Object.keys(extractedActions).length >= 1) {
          actions[key] = extractedActions;
        }
        if (Object.keys(extractedReducer).length >= 1) {
          reducer[key] = extractedReducer;
        }
      } else {
        // Add to state
        state[key] = value;
      }

      return [state, actions, reducer];
    },
    [{}, {}, {}],
  );

export default extractInternals;
