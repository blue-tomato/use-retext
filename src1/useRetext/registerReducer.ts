import mitt from 'mitt';
import { mapValuesDeep } from '../helpers';
import { ReducerFunction } from '../Store';

interface OnCallEvent {
  key: string;
  scope: string;
  reducerFunction: ReducerFunction;
  payload?: unknown;
}

export default (reducer: object, emitter: mitt.Emitter, onCall: (event: OnCallEvent) => void) => {
  mapValuesDeep(reducer, ({ key, value, scope: reducerScope }) => {
    // Reducer listen to emitter
    emitter.on(key, ({ scope: emitterScope, payload }) => {
      // Ensure that the reducer-scope starts at least with the emitter-scope
      if (!reducerScope.startsWith(emitterScope)) return;

      onCall({
        key,
        scope: reducerScope,
        reducerFunction: value as ReducerFunction,
        payload,
      });
    });
  });
};
