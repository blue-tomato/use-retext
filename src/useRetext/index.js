import { useMemo, useState } from 'react';
import { get } from 'lodash-es';
import { assert } from '../helpers';
import createDisptach from './createDispatch';
import connectReducer from './connectReducer';

const mapReducer = (reducer, setState) =>
  Object.entries(reducer).reduce(
    (acc, [key, value]) => ({
      ...acc,
      [key]:
        typeof value === 'object' ? mapReducer(value, setState) : payload => setState(state => value(state, payload)),
    }),
    {},
  );

export default store => {
  assert(typeof store !== 'object', 'Store is not an object');

  const { state: initialState, action, reducer } = store;

  assert(
    typeof initialState !== 'object' || typeof action !== 'object' || typeof reducer !== 'object',
    'State, action or reducer are not an object',
  );

  const [state, setState] = useState(initialState);
  const [disptach, emitter] = createDisptach(action);

  connectReducer(reducer, emitter, ({ scope, reducer: reducerFunction, payload }) => {
    // reducerFunction(get(state, scope), payload);
  });

  const dispatch = useMemo(() => mapReducer(reducer, setState), [reducer]);

  return { state, dispatch };
};
