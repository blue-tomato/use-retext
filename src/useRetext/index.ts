import { useState } from 'react';
import get from 'just-safe-get';
import set from 'just-safe-set';
import { assert } from '../helpers';
import createDispatch from './createDispatch';
import connectReducer from './connectReducer';
import { Store } from '../Store';

export default (store: Store) => {
  assert(typeof store !== 'object', 'Store is not an object');

  const { state: initialState, action, reducer } = store;

  assert(
    typeof initialState !== 'object' || typeof action !== 'object' || typeof reducer !== 'object',
    'State, action or reducer are not an object',
  );

  const [state, setState] = useState(initialState);
  const [dispatch, emitter] = createDispatch(action);

  connectReducer(reducer, emitter, ({ scope, reducer: reducerFunction, payload }) => {
    setState((currentState: any) => {
      const scopedState = scope ? get(currentState, scope) : currentState;
      const result = reducerFunction(scopedState, payload);
      let newState = { ...currentState };

      if (scope) {
        set(newState, scope, { ...scopedState, ...result });
      } else {
        newState = { ...scopedState, ...result };
      }

      return newState;
    });
  });

  return { state, dispatch };
};
