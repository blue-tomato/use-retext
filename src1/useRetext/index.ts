import { useState } from 'react';
import get from 'just-safe-get';
import set from 'just-safe-set';
import typeOf from 'just-typeof';
import { assert } from '../helpers';
import { Store } from '../Store';
import extractInternals from './extractInternals';
import createDispatch from './createDispatch';
import registerReducer from './registerReducer';

export type SelectorFunction = (store: Store) => Store;

export default (store: Store, selectorFunction?: SelectorFunction) => {
  assert(typeOf(store) !== 'object', 'Store is not an object');

  const selectedStore = selectorFunction ? selectorFunction(store) : store;
  const [initialState, actions, reducer] = extractInternals(selectedStore);
  const [state, setState] = useState(initialState);
  const [dispatch, emitter] = createDispatch(actions);

  registerReducer(reducer, emitter, ({ reducerFunction, scope, payload }) => {
    setState((currentState: object) => {
      const scopedState = (scope ? get(currentState, scope) : currentState) as object;
      const result = reducerFunction(scopedState, payload);
      let newState = { ...currentState };

      if (scope) {
        set(newState, scope, { ...scopedState, ...result });
      } else {
        newState = { ...scopedState, ...result };
      }

      return newState;
    });
  });

  return { state, dispatch };
};
