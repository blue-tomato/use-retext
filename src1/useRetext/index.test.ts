import { act, renderHook } from '@testing-library/react-hooks';
import useRetext, { action, reducer } from '.';

const getHook = () =>
  renderHook(() =>
    useRetext({
      count: 1,
      increment: action(),
      setCount: action(),
      _actions: {
        increment: 0,
        setCount: 0,
      },
      _reducer: {
        increment: (state: { count: number }) => ({ count: state.count + 1 }),
        setCount: (state: { count: number }, count: number) => ({ count }),
      },
      sideMenu: {
        isOpen: false,
        maxItems: 100,
        _actions: {
          toggle: 0,
          increment: 0,
        },
        _reducer: {
          toggle: (state: { isOpen: boolean }) => ({ isOpen: !state.isOpen }),
          increment: (state: { maxItems: number }) => ({ maxItems: state.maxItems + 1 }),
        },
        child: {
          isExpanded: false,
          _reducer: {
            toggle: (state: { isExpanded: boolean }) => ({ isExpanded: !state.isExpanded }),
          },
        },
      },
    }),
  );

it('changes the state', () => {
  const { result } = getHook();
  const { dispatch } = result.current as any;

  expect(result.current).toMatchSnapshot();

  act(() => {
    dispatch.setCount(10);
    dispatch.increment();
    dispatch.sideMenu.toggle();
    dispatch.sideMenu.increment();
  });

  expect(result.current).toMatchSnapshot();
});
