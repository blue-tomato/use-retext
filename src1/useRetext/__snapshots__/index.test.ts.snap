// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changes the state 1`] = `
Object {
  "dispatch": Object {
    "increment": [Function],
    "setCount": [Function],
    "sideMenu": Object {
      "increment": [Function],
      "toggle": [Function],
    },
  },
  "state": Object {
    "count": 1,
    "sideMenu": Object {
      "child": Object {
        "isExpanded": false,
      },
      "isOpen": false,
      "maxItems": 100,
    },
  },
}
`;

exports[`changes the state 2`] = `
Object {
  "dispatch": Object {
    "increment": [Function],
    "setCount": [Function],
    "sideMenu": Object {
      "increment": [Function],
      "toggle": [Function],
    },
  },
  "state": Object {
    "count": 11,
    "sideMenu": Object {
      "child": Object {
        "isExpanded": true,
      },
      "isOpen": true,
      "maxItems": 102,
    },
  },
}
`;
